src/daft-local-execution/src/sinks/grouped_aggregate.rs

```diff
@@ -0,0 +1,403 @@
+                let aggregated =
+                    MicroPartition::concat(unaggregated.iter().chain(std::iter::once(&p)))?.agg(
+                        params.partial_agg_exprs.as_slice(),
+                        params.group_by.as_slice(),
+                    )?;
+                state.partially_aggregated.push(aggregated);
+                state.unaggregated_size = 0;
+            } else {
+                state.unaggregated_size += p.len();
+                state.unaggregated.push(p);
+            }
+        }
+        Ok(())
+    }
+
+    fn execute_partition_only(
+        inner_states: &mut [Option<SinglePartitionAggregateState>],
+        input: Arc<MicroPartition>,
+        params: &GroupedAggregateParams,
+    ) -> DaftResult<()> {
+        let partitioned =
+            input.partition_by_hash(params.group_by.as_slice(), inner_states.len())?;
+        for (p, state) in partitioned.into_iter().zip(inner_states.iter_mut()) {
+            let state = state.get_or_insert_with(|| SinglePartitionAggregateState {
+                partially_aggregated: vec![],
+                unaggregated: vec![],
+                unaggregated_size: 0,
+            });
+            state.unaggregated_size += p.len();
+            state.unaggregated.push(p);
+        }
+        Ok(())
+    }
+}
+
+struct SinglePartitionAggregateState {
+    partially_aggregated: Vec<MicroPartition>,
+    unaggregated: Vec<MicroPartition>,
+    unaggregated_size: usize,
+}
+
+enum GroupedAggregateState {
+    Accumulating {
+        inner_states: Vec<Option<SinglePartitionAggregateState>>,
+        strategy: Option<AggStrategy>,
+    },
+    Done,
+}
+
+impl GroupedAggregateState {
+    // This is the threshold for when we should aggregate any accumulated unaggregated partitions
+    const PARTIAL_AGG_THRESHOLD: usize = 10_000;
+    // This is the threshold that indicate high cardinality grouping in the input data, and we should use partitioning first strategy.
+    // It is the ratio between the number of groups vs the number of rows.
+    // For example if the input data had 100 rows and 80 groups or above, that's considered high cardinality.
+    const HIGH_CARDINALITY_THRESHOLD_RATIO: f64 = 0.8;
+
+    fn new(num_partitions: usize) -> Self {
+        let inner_states = (0..num_partitions).map(|_| None).collect();
+        Self::Accumulating {
+            inner_states,
+            strategy: None,
+        }
+    }
+
+    fn push(
+        &mut self,
+        input: Arc<MicroPartition>,
+        params: &GroupedAggregateParams,
+    ) -> DaftResult<()> {
+        if let Self::Accumulating {
+            ref mut inner_states,
+            strategy,
+        } = self
+        {
+            // If we have determined a strategy, execute it.
+            if let Some(strategy) = strategy {
+                strategy.execute_strategy(inner_states, input, params)?;
+            }
+            // Otherwise, determine a strategy based on the input data.
+            else {
```

Comment:
So an issue I see here is that this decision is determined at a per partition basis rather than globally.

This means that as a morsel goes into each core, the decision is made independently of one another so you can have multiple different strategies at the same time.

I think we can instead use a Mutex to protect the strategy (shared across all partition) and then we can cache it in the state after we read it once since it's not going to change.

Reply:
Makes sense, implemented. Tested and found no regressions